functions/AWSECSWrangleASG.js

```diff
@@ -0,0 +1,174 @@
+           console.log(`Container instance: ${containerInstance.ec2InstanceId} has ${usedResource} resource(s) used and ${freeResource} resource(s) available`);
+
+           // If this instance has more resources available then the last highest, make this instance the highest
+           if (freeResource > instanceWithMostResources.FreeResources) {
+                instanceWithMostResources.FreeResources = freeResource;
+                instanceWithMostResources.UsedResources = usedResource;
+                instanceWithMostResources.ID = containerInstance.ec2InstanceId;
+                instanceWithMostResources.ARN = containerInstance.containerInstanceArn;
+           }
+        }
+
+        console.log(`There are ${totalAvailableResources} resource(s) available across the ASG`);
+        console.log(`The least used instance is ${instanceWithMostResources.ID}, it is using ${instanceWithMostResources.UsedResources} resource(s)`);
+
+        // Calculate whether to wrangle or not
+        const freeResourcesAfterWrangle = (totalAvailableResources - instanceWithMostResources.FreeResources) - (instanceWithMostResources.UsedResources);
+        if ( freeResourcesAfterWrangle > 0) {
+            console.log(`We will wrangle, as we have ${freeResourcesAfterWrangle} resource(s) available after this wrangle. This begins by draining the instance ${instanceWithMostResources.ID}`);
+            // We will wrangle, drain the instance in question
+            const drainingInstance = await ecsclient.send(
+                new UpdateContainerInstancesStateCommand({
+                    cluster: params.cluster,
+                    containerInstances: [].concat(instanceWithMostResources.ARN),
+                    status: "DRAINING"
+                }))
+
+            // Wait until the instance has 0 running tasks     
+            let drainingInstanceTimeoutCounter = 0;
+            let drainingInstanceRunningTasks = 1;
+            do {
+                // The division by 5 here is a throttling timeout pattern; to ensure that the ecsclient is not flooded with requests
+                await new Promise(x => setTimeout(x, (drainingInstanceTimeoutSeconds / 5) * 1000));
+                const drainingInstanceUpdate = await ecsclient.send(
+                    new DescribeContainerInstancesCommand({
+                        cluster: params.cluster,
+                        containerInstances: [].concat(instanceWithMostResources.ARN)
+                    }))
+
+                drainingInstanceRunningTasks = parseInt(drainingInstanceUpdate.containerInstances.runningTasksCount);
+
+                drainingInstanceTimeoutCounter++;
+
+                // Another part of throttling timeout pattern, 5 + 1 will trip the timeout and throw an error
+                if (drainingInstanceTimeoutCounter == 6) {
+                    throw `Tried to wrangle but instance did not drain within timeout of ${drainingInstanceTimeoutSeconds} seconds`
```

Comment:
Considering throwing a typed error instead of just a string - [reference](https://github.com/goldbergyoni/nodebestpractices#-22-use-only-the-built-in-error-object)